From a57db8e46974abc5b942ba2f617761da90a13eff Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Sat, 2 Jul 2016 12:45:39 +0200
Subject: [PATCH] iplink_geneve: add tos configuration at link creation

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1339178
Upstream Status: iproute2.git commit f4739b2ee780e
Conflicts: Context changes due to already updated kernel headers.

commit f4739b2ee780e8bba51945bf9f37d3459b67102f
Author: John W. Linville <linville@tuxdriver.com>
Date:   Mon Jun 15 14:37:16 2015 -0400

    iplink_geneve: add tos configuration at link creation

    Signed-off-by: John W. Linville <linville@tuxdriver.com>
---
 ip/iplink_geneve.c    | 26 +++++++++++++++++++++++++-
 man/man8/ip-link.8.in |  6 ++++++
 2 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/ip/iplink_geneve.c b/ip/iplink_geneve.c
index dff08a6..331240a 100644
--- a/ip/iplink_geneve.c
+++ b/ip/iplink_geneve.c
@@ -11,16 +11,18 @@
 
 #include <stdio.h>
 
+#include "rt_names.h"
 #include "utils.h"
 #include "ip_common.h"
 
 static void print_explain(FILE *f)
 {
 	fprintf(f, "Usage: ... geneve id VNI remote ADDR\n");
-	fprintf(f, "                 [ ttl TTL ]\n");
+	fprintf(f, "                 [ ttl TTL ] [ tos TOS ]\n");
 	fprintf(f, "\n");
 	fprintf(f, "Where: VNI  := 0-16777215\n");
 	fprintf(f, "       ADDR := IP_ADDRESS\n");
+	fprintf(f, "       TOS  := { NUMBER | inherit }\n");
 	fprintf(f, "       TTL  := { 1..255 | inherit }\n");
 }
 
@@ -37,6 +39,7 @@ static int geneve_parse_opt(struct link_util *lu, int argc, char **argv,
 	__u32 daddr = 0;
 	struct in6_addr daddr6 = IN6ADDR_ANY_INIT;
 	__u8 ttl = 0;
+	__u8 tos = 0;
 
 	while (argc > 0) {
 		if (!matches(*argv, "id") ||
@@ -66,6 +69,17 @@ static int geneve_parse_opt(struct link_util *lu, int argc, char **argv,
 					invarg("TTL must be <= 255", *argv);
 				ttl = uval;
 			}
+		} else if (!matches(*argv, "tos") ||
+			   !matches(*argv, "dsfield")) {
+			__u32 uval;
+
+			NEXT_ARG();
+			if (strcmp(*argv, "inherit") != 0) {
+				if (rtnl_dsfield_a2n(&uval, *argv))
+					invarg("bad TOS value", *argv);
+				tos = uval;
+			} else
+				tos = 1;
 		} else if (matches(*argv, "help") == 0) {
 			explain();
 			return -1;
@@ -95,6 +109,7 @@ static int geneve_parse_opt(struct link_util *lu, int argc, char **argv,
 	if (daddr)
 		addattr_l(n, 1024, IFLA_GENEVE_REMOTE, &daddr, 4);
 	addattr8(n, 1024, IFLA_GENEVE_TTL, ttl);
+	addattr8(n, 1024, IFLA_GENEVE_TOS, tos);
 
 	return 0;
 }
@@ -103,6 +118,7 @@ static void geneve_print_opt(struct link_util *lu, FILE *f, struct rtattr *tb[])
 {
 	__u32 vni;
 	char s1[1024];
+	__u8 tos;
 
 	if (!tb)
 		return;
@@ -126,6 +142,14 @@ static void geneve_print_opt(struct link_util *lu, FILE *f, struct rtattr *tb[])
 		if (ttl)
 			fprintf(f, "ttl %d ", ttl);
 	}
+
+	if (tb[IFLA_GENEVE_TOS] &&
+	    (tos = rta_getattr_u8(tb[IFLA_GENEVE_TOS]))) {
+		if (tos == 1)
+			fprintf(f, "tos inherit ");
+		else
+			fprintf(f, "tos %#x ", tos);
+	}
 }
 
 static void geneve_print_help(struct link_util *lu, int argc, char **argv,
diff --git a/man/man8/ip-link.8.in b/man/man8/ip-link.8.in
index 284ee13..31a6190 100644
--- a/man/man8/ip-link.8.in
+++ b/man/man8/ip-link.8.in
@@ -672,6 +672,8 @@ the following additional arguments are supported:
 .BI type " geneve " id " ID " remote " IPADDR"
 .R " [ "
 .BI ttl " TTL "
+.R " ] [ "
+.BI tos " TOS "
 .R " ]"
 
 .in +8
@@ -687,6 +689,10 @@ the following additional arguments are supported:
 .BI ttl " TTL"
 - specifies the TTL value to use in outgoing packets.
 
+.sp
+.BI tos " TOS"
+- specifies the TOS value to use in outgoing packets.
+
 .in -8
 
 .TP
-- 
1.8.3.1

